### ISSUE TITLE
Getting "NUnit.Engine.NUnitEngineException : Unable to acquire remote process agent" when using latest official 3.15.5 version against NUnit 4.0.1 under .NET 6, but not when installing into a directory path not containing spaces (+ related .NET 8 problem)

### ISSUE THREAD

**Issue by mgroetan** [nunit/nunit-console#1384]
This in nature seems similar to #436.
Even more so looks very similar to #1206, which Charlie P says was fixed in 3.16.0.
Keeping in mind the current mess with the 3.15.x version range being published AFTER the 3.16.x range, is there a slight possibility that 3.15.x was in fact built off the wrong baseline, and indeed lacks this fix?

**In short, the exception in the issue title occurs when the NUnit Console executable is located in a directory path that contains spaces under Windows 10**, e.g.  "C:\Program Files (x86)\NUnit.org\nunit-console".
Whereas if run from a directory path that does NOT contain any spaces, it does work.

In addition, running the same command against a .NET **8** test assembly, it fails the same way _regardless_ of where the console executable is located (more details to follow further down).

**The command line being used is as simple as it can get:**
`nunit3-console.exe Redacted.Tests.dll`

**Method of installation:**
msi

**Version of the NUnit Engine/Console:**
3.15.5

**All versions of the NUnit Framework in use:**
4.0.1
NUnit3TestAdapter: 4.5.0

**Whether the behaviour is any different when the tests are run using the --inprocess flag:**
Yes, then it fails catastrophically, unable to load the "System.Runtime" assembly.

**Summary of how various NUnit Console versions behave (with more detailed logs to follow):**
.NET 6:
3.16.1 does NOT work with NUnit 4.x - unable to find any tests (expected, as it lacks support)
3.15.5 does NOT work when run from a directory containing spaces (the default installation location) - unable to create test agent
3.15.5 does work when run from a directory NOT containing spaces
3.17.0-dev.3 does NOT work when run from a directory containing spaces (the default installation location) - unable to create test agent
3.17.0-dev.3 does work when run from a directory NOT containing spaces

.NET 8:
3.16.1 does NOT work with NUnit 4.x - unable to recognize .NET 8 (expected, as it lacks support)
3.15.5 does NOT work when run from a directory containing spaces (the default installation location) - unable to create test agent
3.15.5 does NOT work when run from a directory NOT containing spaces - unable to recognize .NET 8 (expected, as it lacks support)
3.17.0-dev.3 does NOT work when run from a directory containing spaces (the default installation location) - unable to create test agent
3.17.0-dev.3 does NOT work when run from a directory NOT containing spaces - unable to create test agent

**Log details:**
3.15.5 against .NET 6 - default installation directory:
```
C:\Git\Redacted.Tests\bin\Debug\net6.0>"C:\Program Files (x86)\NUnit.org\nunit-console\nunit3-console.exe" Redacted.Tests.dll
NUnit Console 3.15.5 (Release)
Copyright (c) 2022 Charlie Poole, Rob Prouse
onsdag 17. januar 2024 11:07:50

Runtime Environment
   OS Version: Microsoft Windows NT 6.2.9200.0
   Runtime: .NET Framework CLR v4.0.30319.42000

Test Files
    Redacted.Tests.dll


Unhandled Exception: NUnit.Engine.NUnitEngineException: Remote test agent exited with non-zero exit code 1
   at NUnit.Engine.Services.TestAgency.OnAgentExit(Process process, Guid agentId)
   at NUnit.Engine.Services.TestAgency.<>c__DisplayClass13_0.<GetAgent>b__0(Object sender, EventArgs e)
   at System.Diagnostics.Process.OnExited()
   at System.Diagnostics.Process.RaiseOnExited()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading._ThreadPoolWaitOrTimerCallback.PerformWaitOrTimerCallback(Object state, Boolean timedOut)

Errors, Failures and Warnings

1) Error :
NUnit.Engine.NUnitEngineException : Unable to acquire remote process agent
--NUnitEngineException
Unable to acquire remote process agent
   at NUnit.Engine.Runners.ProcessRunner.CreateAgentAndRunnerIfNeeded()
   at NUnit.Engine.Runners.ProcessRunner.RunTests(ITestEventListener listener, TestFilter filter)

Test Run Summary
  Overall result: Failed
```

3.15.5 against .NET 6 - custom installation directory without spaces:
```
C:\Git\Redacted.Tests\bin\Debug\net6.0>"C:\Temp\nunit\3.15.5\nunit-console\nunit3-console.exe" Redacted.Tests.dll
NUnit Console 3.15.5 (Release)
Copyright (c) 2022 Charlie Poole, Rob Prouse
onsdag 17. januar 2024 11:09:54

Runtime Environment
   OS Version: Microsoft Windows NT 6.2.9200.0
   Runtime: .NET Framework CLR v4.0.30319.42000

Test Files
    Redacted.Tests.dll

Run Settings
    DisposeRunners: True
    WorkDirectory: C:\Git\Redacted.Tests\bin\Debug\net6.0
    ImageRuntimeVersion: 4.0.30319
    ImageTargetFrameworkName: .NETCoreApp,Version=v6.0
    ImageRequiresX86: False
    ImageRequiresDefaultAppDomainAssemblyResolver: False
    TargetRuntimeFramework: netcore-6.0
    NumberOfTestWorkers: 16

Test Run Summary
  Overall result: Passed
  Test Count: 110, Passed: 110, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 0
```

3.17.0-dev.3 against .NET 6 - default installation directory (this time even with debug tracing turned on):
```
C:\Git\Redacted.Tests\bin\Debug\net6.0>"C:\Program Files (x86)\NUnit.org\nunit-console\nunit3-console.exe" Redacted.Tests.dll --trace=Debug
NUnit Console 3.17.0-dev00003+685c5b542b5e9ba632c905f0bd514a773d9758af (Release)
Copyright (c) 2022 Charlie Poole, Rob Prouse
onsdag 17. januar 2024 13:00:59

Runtime Environment
   OS Version: Microsoft Windows NT 6.2.9200.0
   Runtime: .NET Framework CLR v4.0.30319.42000

Test Files
    Redacted.Tests.dll


Errors, Failures and Warnings

1) Error :
NUnit.Engine.NUnitEngineException : Remote test agent exited with non-zero exit code 1
--NUnitEngineException
Remote test agent exited with non-zero exit code 1
   at NUnit.Engine.Services.TestAgency.OnAgentExit(Process process, Guid agentId)
   at NUnit.Engine.Services.TestAgency.<>c__DisplayClass13_0.<GetAgent>b__0(Object sender, EventArgs e)
   at System.Diagnostics.Process.OnExited()
   at System.Diagnostics.Process.RaiseOnExited()
   at System.Diagnostics.Process.get_HasExited()
   at NUnit.Engine.Services.TestAgency.GetAgent(TestPackage package)
   at NUnit.Engine.Runners.ProcessRunner.CreateAgentAndRunnerIfNeeded()
   at NUnit.Engine.Runners.ProcessRunner.RunTests(ITestEventListener listener, TestFilter filter)

Test Run Summary
  Overall result: Failed
```

3.17.0-dev.3 against .NET 6 - custom installation directory without spaces (this time even with debug tracing turned on):
```
C:\Git\Redacted.Tests\bin\Debug\net6.0>"C:\Temp\nunit\3.17.0-dev.3\nunit-console\nunit3-console.exe" Redacted.Tests.dll --trace=Debug
NUnit Console 3.17.0-dev00003+685c5b542b5e9ba632c905f0bd514a773d9758af (Release)
Copyright (c) 2022 Charlie Poole, Rob Prouse
onsdag 17. januar 2024 13:02:02

Runtime Environment
   OS Version: Microsoft Windows NT 6.2.9200.0
   Runtime: .NET Framework CLR v4.0.30319.42000

Test Files
    Redacted.Tests.dll

Run Settings
    DisposeRunners: True
    InternalTraceLevel: Debug
    WorkDirectory: C:\Git\Redacted.Tests\bin\Debug\net6.0
    ImageRuntimeVersion: 4.0.30319
    ImageTargetFrameworkName: .NETCoreApp,Version=v6.0
    ImageRequiresX86: False
    ImageRequiresDefaultAppDomainAssemblyResolver: False
    TargetRuntimeFramework: netcore-6.0
    NumberOfTestWorkers: 16

Test Run Summary
  Overall result: Passed
  Test Count: 110, Passed: 110, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 0
```

3.17.0-dev.3 against .NET 8 - default installation directory (this time even with debug tracing turned on):
```
C:\Git\Other.Redacted.Tests\bin\Debug\net8.0>"C:\Program Files (x86)\NUnit.org\nunit-console\nunit3-console.exe" Other.Redacted.Tests.dll --debug=Trace
NUnit Console 3.17.0-dev00003+685c5b542b5e9ba632c905f0bd514a773d9758af (Release)
Copyright (c) 2022 Charlie Poole, Rob Prouse
onsdag 17. januar 2024 11:43:11

Runtime Environment
   OS Version: Microsoft Windows NT 6.2.9200.0
   Runtime: .NET Framework CLR v4.0.30319.42000

Test Files
    Other.Redacted.Tests.dll


Unhandled Exception: NUnit.Engine.NUnitEngineException: Remote test agent exited with non-zero exit code 1
   at NUnit.Engine.Services.TestAgency.OnAgentExit(Process process, Guid agentId)
   at NUnit.Engine.Services.TestAgency.<>c__DisplayClass13_0.<GetAgent>b__0(Object sender, EventArgs e)
   at System.Diagnostics.Process.OnExited()
   at System.Diagnostics.Process.RaiseOnExited()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading._ThreadPoolWaitOrTimerCallback.PerformWaitOrTimerCallback(Object state, Boolean timedOut)

Errors, Failures and Warnings

1) Error :
NUnit.Engine.NUnitEngineException : Unable to acquire remote process agent
--NUnitEngineException
Unable to acquire remote process agent
   at NUnit.Engine.Runners.ProcessRunner.CreateAgentAndRunnerIfNeeded()
   at NUnit.Engine.Runners.ProcessRunner.RunTests(ITestEventListener listener, TestFilter filter)

Test Run Summary
  Overall result: Failed
```

3.17.0-dev.3 against .NET 8 - custom installation directory without spaces (this time even with debug tracing turned on):
```
C:\Git\Other.Redacted.Tests\bin\Debug\net8.0>"C:\Temp\nunit\3.17.0-dev.3\nunit-console\nunit3-console.exe" Other.Redacted.Tests.dll --debug=Trace
NUnit Console 3.17.0-dev00003+685c5b542b5e9ba632c905f0bd514a773d9758af (Release)
Copyright (c) 2022 Charlie Poole, Rob Prouse
onsdag 17. januar 2024 13:32:21

Runtime Environment
   OS Version: Microsoft Windows NT 6.2.9200.0
   Runtime: .NET Framework CLR v4.0.30319.42000

Test Files
    Other.Redacted.Tests.dll


Errors, Failures and Warnings

1) Error :
NUnit.Engine.NUnitEngineException : Remote test agent exited with non-zero exit code 1
--NUnitEngineException
Remote test agent exited with non-zero exit code 1
   at NUnit.Engine.Services.TestAgency.OnAgentExit(Process process, Guid agentId)
   at NUnit.Engine.Services.TestAgency.<>c__DisplayClass13_0.<GetAgent>b__0(Object sender, EventArgs e)
   at System.Diagnostics.Process.OnExited()
   at System.Diagnostics.Process.RaiseOnExited()
   at System.Diagnostics.Process.get_HasExited()
   at NUnit.Engine.Services.TestAgency.GetAgent(TestPackage package)
   at NUnit.Engine.Runners.ProcessRunner.CreateAgentAndRunnerIfNeeded()
   at NUnit.Engine.Runners.ProcessRunner.RunTests(ITestEventListener listener, TestFilter filter)

Test Run Summary
  Overall result: Failed
```

For now, we've been able to restore our successful test runs by temporarily downgrading to NUnit 3.14.0 again. Even though there seems to be a possible workaround by installing into a "space-less" directory path, we've chosen not to go down that route for now.

Are there possibly some related changes needed to the "NUnit3TestAdapter"? Will there be an "NUnit**4**TestAdapter" and "nunit**4**-console.exe" at some point?

**Comment by stevenaw** [comment]
Thanks for the report and thorough up front investigation @mgroetan 

I agree that it does seem related to #1206 . The team will often backport bugfixes to previous releases if the issue is known to exist there, especially if there's uncertainty around the timing of the next semver-minor release, however it looks like the commit which fixed this in 3.16 was not backported to a 3.15.x release. This may explain what you're seeing (https://github.com/nunit/nunit-console/pull/1207/commits/1814fda19bffbbd77f3203ce5d3b9fad6f82cb21)

The 3.16 release had a few issues and so we've recently released a stable 3.17 which _also_ wouldn't include this as we based 3.17 off the 3.15 branch as it was known to be more stable.

I think there's a good chance that applying the above linked commit to 3.17 may fix the issue, but we'll want to be careful applying it verbatim just in case it was the cause of some of the issues we saw in 3.16

**Comment by stevenaw** [comment]
I'm labeling this as needs confirmation as I haven't actually run any code to definitively repro this.
